### ISSUE TITLE
Packages: Move `animation` to new structure

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#6042]
## Feature Description

* [ ] Move code from `/assets/src/animation` to `/packages/animation/src/`
* [ ] Update imports where necessary.  
  Use `@web-stories-wp/...` imports for anything that is already a package, relative imports for everything else.
* [ ] Create `/packages/animation/package.json` referencing its dependencies
* [ ] Create `/packages/animation/README.md` with short documentation.

### Blockers ⚠️ 

Right now `assets/src/animation` has a one dependency on the editor:

* `assets/src/edit-story/types` - for story element prop types

At the same time, the editor includes a bunch of code from `animation`.

We first need to resolve this circular dependency before `animation` can be  become its own package.

Possible solution:

* Create a new `@web-stories-wp/elements` package that contains elements-related utils, JSDoc types, and prop types. Elements themselves could be in the same package, or a new `element-library` package.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by swissspidy** [comment]
@littlemilkstudio Heya 👋 since you worked quite a bit on animations, any thoughts on how we could best avoid the circular dependencies between animation <-> editor? Left some ideas in the description above already.

**Comment by maxyinger** [comment]
@swissspidy Hmph. yea. Looks like `assets/src/edit-story/output` isn't really much of a dep since it's just used for storybook. Since it's storybook specific, and they're small, I don't think there's any harm in copying them like you mentioned.

As for hoisting constants, data structure definition & pure fn utils into a separate package I think that's a good move. It kinda seems to me that the editor has two separate things it's doing rn. Defining what a story is schema/constant wise then another layer is interpreting and interacting with that schema.

I think separating the two responsibilities into separate packages could be nice. Would allow us to create alternate interpretation layers in the future that are only dependent on the schema, constant & utils (a schema -> svg interpreter comes to mind).

I'm pretty unopinionated as to how we separate out constants, data structure definition & pure fns, but agree that they should live in a separate package that the editor and the animation package both pull from.

tldr: Yea I think your suggestions make sense

**Comment by maxyinger** [comment]
Also, If we were to ever do package by package transitions to TS, the story schema package would be a super nice place to start that would play really well with TS

**Comment by spacedmonkey** [comment]
I wonder if @barklund has any thoughts on this matter.

**Comment by maxyinger** [comment]
> I wonder if @barklund has any thoughts on this matter.

same.

**Comment by barklund** [comment]
I'm not fully seeing the consequences of the suggested ideas, but don't think I have strong ideas one way or another.

I'm a fan of DRY, but even more that of well-organized code, so if we have to sacrifice the former for the latter, I'm more than okay with that.

The initiative here and ideas all sound very promising (although a bit daunting), and I commend any effort aiming for better organization! 👏  🎉

**Comment by spacedmonkey** [comment]
We do not only need to solve these interdepencies for this package, but for Dashboard ( #6052 ) as well.

Having a quick look at the dashboard codebase, they are some similar interdepencies. 

![Screenshot 2021-02-12 at 11 29 14](https://user-images.githubusercontent.com/237508/107763014-cb7b3900-6d25-11eb-92ff-0dab06d25876.png)

So common dependencies that jump out here. 

`assets/src/edit-stories/units`
The provider is used in dashboard. Maybe instead of just moving `edit-story/units/dimensions`, provider and all could be moved to new package. 

`assets/src/edit-story/types` Also used.
`assets/src/edit-story/constants` Also used.
`assets/src/edit-story/elements/media` - for the visualImage util

I recommend the following packages.
 
### @web-stories-wp/elements
Move all the elements to package. This would be similar to Gutenberg's [block-library](https://github.com/WordPress/gutenberg/tree/master/packages/block-library) package. 

### @web-stories-wp/types
Move type definitions to its own package. 

### @web-stories-wp/constants
Move constants definitions to its own package. 

### @web-stories-wp/utils
Move all existing utils from edit stories ( and maybe even dashboard) to new package. Move `Boilerplate` and `CustomCSS` to new utils package. 

### @web-stories-wp/units
Move provider and  dimensions. I know that this relays on design system for now, but we should handle this once we make the design system it's own package. 

I think with these new packages, it will cover around 90-95% dependencies I have documented here. I know it will not be perfect, but we can deal with outliers on a case by case basis.

**Comment by swissspidy** [comment]
I didn't anticipate this becoming the catch-all thread for discussing package structure 😄 

Y'all are sharing some very good thoughts here though.

There are many interdependencies throughout the codebase, but we shouldn't just move all the least common denominators into their own packages. We need to be smart about it and do it well.

To quickly go over the suggestions:

* `@web-stories-wp/elements`: I was thinking about this one as well, but these have quite a few dependencies that need to be sorted out. So not a simple task.
* `@web-stories-wp/types`: types should live next to the actual code, not in a separate package. e.g. element types should live in the elements package.
* `@web-stories-wp/constants`: same here, don't think this makes sense as a package. e.g. element constants should live in the elements package.
* `@web-stories-wp/utils`: same here, utils should live closer to where they are needed. I'd want to avoid a catch-all package for random utils.
* >  Move `Boilerplate` and `CustomCSS` to new utils package.    
  
  Those two examples would actually make more sense as part of a `@web-stories-wp/serialization` package that deals with AMP story output. But this one has some dependencies on its own as well.  
* `@web-stories-wp/units`: still seems very lightweight, and perhaps needs a better name, but could work
* > I know that this relies on design system for now, but we should handle this once we make the design system it's own package.  
  
  If you are referring to the dependency on `createContext`, that would actually make sense to live in its own `@web-stories-wp/context` package
* >  It kinda seems to me that the editor has two separate things it's doing rn. Defining what a story is schema/constant wise then another layer is interpreting and interacting with that schema.
  >
  >I think separating the two responsibilities into separate packages could be nice. Would allow us to create alternate interpretation layers in the future that are only dependent on the schema, constant & utils (a schema -> svg interpreter comes to mind).  
  
  Yup, makes sense. Something like `@web-stories-wp/stories` or so for holding the schema could make sense.

**Comment by BrittanyIRL** [comment]
This is quite the feat! 

I don't have a lot of opinions on how this is done other than I agree with Pascal that types, constants and utils make sense to be kept close to where they are used not isolated as packages.

**Comment by spacedmonkey** [comment]
As I was writing I didn't really like having types / constants as packages, they are definitions of types and values, not functions. Maybe we need to find another way to share code. 

But I don't agree that these types should live next the code in some case. Defining elements, types and consts are values shared between dashboard and editor, not just defined in the editor. 

It feels like the editor pulling double duty ( similar to what Max said ), define data types / values, rendering and editing. So the dependency is like this. 
data types -> render -> edit
data types -> render -> view in dashboard.  

I would like to see rendering the story as another package, both dashboard and editor need to do this and other parts of the code may do this later like a gutenberg block for example. This is matter of sharing code, I don't care how we do it, just don't like how we are currently doing with cherry picking parts from the editor and gluing them into the dashboard.

**Comment by swissspidy** [comment]
> But I don't agree that these types should live next the code in some case. Defining elements, types and consts are values shared between dashboard and editor, not just defined in the editor.

That's not what I meant. To avoid confusion: If there would be packages like `@web-stories-wp/stories` or `@web-stories-wp/elements`, the types and consts for elements and the story schema would live _there_, _not_ the _editor_.

----

Looks like we're overall trending towards the same thing.

I'll try to write this up in a more formalized way as a small design doc so we can more easily discuss specifics.

**Comment by swissspidy** [comment]
Current WIP document: https://docs.google.com/document/d/1ImRe9wYKWGCAZiq3jro5KLpSJK_NMKXHC6FWqs3nwl0/edit

Based on exploratory work in #6902

**Comment by spacedmonkey** [comment]
Highlighting the current blockers. 

https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/storybookUtils/index.js#L25

This is just used in storybook. Do we need it? Can we just copy this over?

https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/storybookUtils/index.js#L26

This can not be copied over, as CustomCSS uses theme. 

https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/edit-story/output/utils/styles.js#L25

https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/components/provider.js#L40

Not sure what to do about this one. Once edit story is it's own package, then we import these types. Or do we need a new package. 

https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/utils/getGlobalSpace.js#L18

As above, we can import these values from a edit story package it is in place. 

@swissspidy Can you advise here so we can unblock this. We are sooo nearly there.

**Comment by swissspidy** [comment]
> Highlighting the current blockers.
> 
> https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/storybookUtils/index.js#L25
> 
> This is just used in storybook. Do we need it? Can we just copy this over?

Just copy it over 👍 

> https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/storybookUtils/index.js#L26
> 
> This can not be copied over, as CustomCSS uses theme.

This usage can be removed.

> https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/edit-story/output/utils/styles.js#L25
> 
> https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/components/provider.js#L40
> 
> Not sure what to do about this one. Once edit story is its own package, then we import these types. Or do we need a new package.
> 
> https://github.com/google/web-stories-wp/blob/6687928bfc1eee50a07049ed73dcf835b960e56b/assets/src/animation/utils/getGlobalSpace.js#L18
> 
> As above, we can import these values from a edit story package it is in place.

As per the ticket description, these cannot be imported from edit-story, as that would be a circular dependency. The current thought is to create a new `elements` package.
